### PR TITLE
Add custom provider

### PR DESCRIPTION
**Pull request by glethuillier**
**+ Support for custom provider**

To use the new provider, please rename the `.env.template` file in `.env` and set the URL of the custom provider as well as your API key. Then, build and run.

When the custom provider is being used, xpub scan displays `(Data fetched from the custom provider)`. Otherwise, it displays `(Data fetched from the default provider)`.

**+ Differentiation between sent to self `⮂` and sent to sibling `↺`**

Xpub scan is now able to tell the difference between a transaction sent to self (that is: when an address directly sends to itself), and a transaction sent to a sibling (that is: when an address sends to a non-change address that belongs to the same xpub but not to itself).

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -0,0 +1,2 @@
+API_URL=https://provider.com/{coin}/address/{address}
+API_KEY=<your api key>
```
Comment (vbergeron-ledger) [inline .env.template:2]: new line missing

File: src/helpers.ts [hunk 2]
```diff
@@ -37,24 +42,36 @@ function checkXpub(xpub: string) {
...
+    )
+  );
 }
 
 export {
-  checkXpub,
-  getJSON
+  getJSON,
+  checkXpub
 }
```
Comment (vbergeron-ledger) [inline src/helpers.ts:77]: new line missing

File: src/models/operation.ts [hunk 3]
```diff
@@ -15,13 +15,17 @@ class Operation {
     address: string;
     amount: number;
 
-    // self sent (sent to non-change address belonging to same xpub)
+    // self sent (sent to same address)
     self: boolean; 
 
+    // sent to non-change address belonging to same xpub
+    sentToSibling: boolean;
+
```
Comment (vbergeron-ledger) [inline src/models/operation.ts:23]: you could create an enum describing the the operation type

File: src/helpers.ts [hunk 4]
```diff
@@ -37,24 +42,36 @@ function checkXpub(xpub: string) {
   const prefix = xpub.substring(0, 4);
 
   if (prefix === 'xpub') {
-    network.type = BITCOIN_NETWORK;
+    configuration.network = BITCOIN_NETWORK;
   }
   else if (prefix === 'Ltub') {
-    network.type = LITECOIN_NETWORK;
+    configuration.network = LITECOIN_NETWORK;
   }
```
Comment (YBadiss) [inline src/helpers.ts:49]: As you add more coins, you might want to aim for something like

```javascript
const xpubPrefix = xpub.substring(0, 4);
const network = NETWORKS.find( ({ prefix }) => prefix === xpubPrefix );
if (!network) { throw ... }
```

**Review by vbergeron-ledger — APPROVED** [review]

File: src/helpers.ts [hunk 5]
```diff
@@ -3,14 +3,25 @@ import request from "sync-request";
...
 // TODO: properly rework this function
-function getJSON(url: string, attempts = 0) {
-  const res = request('GET', url);
+function getJSON(url: string, APIKey?: string) {
+  let headers = {};
 
-  if (attempts > 5) {
+  if (APIKey !== undefined) {
+    headers = {
+      'X-API-Key': APIKey
```
Comment (YBadiss) [inline src/helpers.ts:18]: While this is meant to be generic, there is no guarantee other provides will use this header name.

**Review by YBadiss — COMMENTED** [review]
I think this is a good effort to have a "standard" approach to other providers but I am not sure it works as different providers will have clearly different interfaces.

It may be less hassle/work to clearly define which providers we support, and explicitly define the expected message schemas.
In that case, adding a a new provider should ideally be as simple as adding a new file (defining the interface, API key handling, etc.) + adding the provider in a supported list.

File: src/helpers.ts (hunk 4 above)
Comment (glethuillier) [reply src/helpers.ts:49]: Yes, I agree that this aspect will have to be refactored in the future.

File: src/helpers.ts (hunk 5 above)
Comment (glethuillier) [reply src/helpers.ts:18]: Indeed. This whole function will have to be refactored in the future (including making it async, handle retries, etc.).

File: src/models/operation.ts (hunk 3 above)
Comment (glethuillier) [reply src/models/operation.ts:25]: Done.

File: src/helpers.ts (hunk 5 above)
Comment (glethuillier) [reply src/helpers.ts:18]: I suggest that we keep that way at the moment, and refactor it completely later. What do you think @YBadiss?

File: .env.template (hunk 1 above)
Comment (glethuillier) [reply .env.template]: Done.

File: src/helpers.ts (hunk 2 above)
Comment (glethuillier) [reply src/helpers.ts]: Done.

**Comment by YBadiss** [comment]
PS: I realise it was not clear, but this PR looks fine to me as it is :)

**Pull request merged**